File: edition/2017/java/Sorts/insertion_sort.java
/**
 * Created by renu.yadav on 20/10/17.
 */
public class insertion_sort {
    /**
     * insertion sort picks up the current element(i) and checks it against its previous element(j = i-1 until j>=0) and place it
     * takes O(n^2) and inplace algo with no extra space
     * @param arr
     * @return
     */

    public static int[] insertionSort(int[] arr) {
        for (int i = 0; i < arr.length; i++) {
            int keyToBeInserted = arr[i];
            int j = i - 1;
            while (j >= 0 && arr[j] > keyToBeInserted) {
                arr[j + 1] = arr[j];
            }
            arr[j + 1] = keyToBeInserted;
        }
        return arr;
    }
}

File: insertionsort/insertionsort.c
#include<stdio.h>
#include<conio.h>
void insertionsort(int arr[],int n)
{int i,j;
int key;
for(i=1;i<n;i++)
{key=arr[i];
j=i-1;
while(j>=0&&arr[j]>key)
{arr[j+1]=arr[j];
j=j-1;
}
arr[j+1]=key;
}
printf("\n");
printf("array after insertion sort");
for(i=0;i<5;i++)
    {
        printf("%d\n",arr[i]);
    }
    printf("\n");}
int main()
{int arr[5],n;
int i;
printf("enter values");
for(i=0;i<5;i++)
{scanf("%d",&arr[i]);}
insertionsort(arr,5);
return 0;
}
